File: tests/cases/check/frozen-rules/python/inputs/baseline/orders/billing.py
from cdec_rules import immutable, no_instantiation, no_side_effects, sealed


@immutable
@sealed
class Receipt:
    def __init__(self, total: float) -> None:
        self.total: float = total


@no_instantiation(allow=["AuditEntry"])
class CheckoutService:
    @no_side_effects
    def subtotal(self, n: int) -> float:
        return float(n)

File: tests/cases/check/frozen-rules/csharp/inputs/baseline/Orders/Billing.cs
using CodeConstraints.Rules;

namespace Orders;

[Immutable]
[Sealed]
public class Receipt
{
    public decimal Total { get; }

    public Receipt(decimal total) { Total = total; }
}

[NoInstantiation(Allow = new[] { "AuditEntry" })]
public class CheckoutService
{
    [NoSideEffects]
    public decimal Subtotal(int n) { return n; }
}
